pkg/tfcli/builders.go

```diff
@@ -0,0 +1,267 @@
+	WithState(tfState []byte) Builder
+}
+
+type RequiresResource interface {
+	WithResourceType(labelType string) Builder
+	WithResourceName(labelName string) Builder
+	WithResourceBody(body []byte) Builder
+	WithHandle(handle string) Builder
+}
+
+type RequiresProvider interface {
+	WithProviderSource(source string) Builder
+	WithProviderVersion(version string) Builder
+	WithProviderConfiguration(conf []byte) Builder
+}
+
+type clientBuilder struct {
+	c *client
+}
+
+func NewClientBuilder() *clientBuilder {
+	c := defaultClient()
+	return &clientBuilder{
+		c: c,
+	}
+}
+
+func (cb clientBuilder) validateNoState() error {
+	if err := cb.c.resource.validate(); err != nil {
+		return err
+	}
+	if err := cb.c.provider.validate(); err != nil {
+		return err
+	}
+	if err := cb.c.logger.validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (cb clientBuilder) Build(ctx context.Context) (Client, error) {
+	if err := cb.validateNoState(); err != nil {
+		return nil, err
+	}
+	if err := cb.c.init(ctx); err != nil {
+		return nil, err
+	}
+	return cb.c, nil
+}
+
+func defaultClient() *client {
+	return &client{
+		state:       &withState{},
+		provider:    providerFromEnv(),
+		resource:    &withResource{},
+		execTimeout: timeoutFromEnv(),
+	}
+}
+
+type withTimeout struct {
+	to time.Duration
+}
+
+func timeoutFromEnv() *withTimeout {
+	result := &withTimeout{}
+	toStr := os.Getenv(envTFExecTimeout)
+	if toStr == "" {
+		return result
+	}
+
+	// silently ignore unparsable timeout errors, we will observe
+	// execution timeout values in logs
+	result.to, _ = time.ParseDuration(toStr)
+	return result
+}
+
+type withLogger struct {
+	log logging.Logger
+}
+
+func (l withLogger) validate() error {
+	if l.log == nil {
+		return errors.New(errValidationNoLogger)
+	}
+	return nil
+}
+
+type withState struct {
+	tfState []byte
+}
+
+func (s withState) validate() error {
+	if len(s.tfState) == 0 {
+		return errors.New(errValidationState)
+	}
+	return nil
+}
+
+// holds values for the Terraform HCL resource block's two labels and body
+type withResource struct {
+	labelType string
+	labelName string
+	body      []byte
+	handle    string
+}
+
+func (r withResource) validate() error {
+	if r.labelName == "" || r.labelType == "" {
+		return errors.Errorf(fmtErrValidationResource, r.labelType, r.labelName)
+	}
+	return nil
+}
+
+func (r withResource) GetAddress() string {
+	return fmt.Sprintf(fmtResourceAddress, r.labelType, r.labelName)
+}
+
+// holds values for the Terraform HCL provider block's source, version and configuration body
+type withProvider struct {
+	source        string
+	version       string
+	configuration []byte
+}
+
+func (p withProvider) validate() error {
+	if p.source == "" || p.version == "" {
+		return errors.Errorf(fmtErrValidationProvider, p.source, p.version)
+	}
+	return nil
+}
+
+func providerFromEnv() *withProvider {
+	p := &withProvider{}
+	if p.source == "" {
+		p.source = os.Getenv(envTFProviderSource)
```

Comment:
Environment variables carry some undesired traits similar to global variables but they also provide some convenience when you have layers of abstractions, so that you can control the behavior at a lower layer without having to flow configuration through intermediate layers. Terraform CLI's certain behavior is exclusively controlled via environment variables and similar to that it was convenient to introduce these two environment variables (`XP_TERRAFORM_PROVIDER_SOURCE` and `XP_TERRAFORM_PROVIDER_VERSION`) that, together with some Terraform configuration via the `TF_CLI_CONFIG_FILE` env. variable control the plugin binary mirroring behavior. 

I would suggest to keep them initially for convenience & development purposes but what's important here is, as we have previously discussed, we need to have the source & version info of the Terraform provider against which we have generated code available at runtime for the `tfcli` library and ultimately for the Terraform CLI.

Reply:
Unfortunately I do have a strong opinion about environment variables. 🤓 I'm all for using them, but I strongly prefer them to all be "loaded" from the environment in one, predictable place - typically in a program's `main()` function. Having `os.Getenv` calls scattered around the codebase makes it difficult to keep track of the various environment variables (and more broadly, configuration sources) that a program uses. For libraries I feel it is best to take configuration as inputs and leave things operational like configuration, logging, etc to the caller.